File: customer-service/src/main/java/com/wks/servicemarketplace/customerservice/adapters/events/DefaultMessagePublisher.java
package com.wks.servicemarketplace.customerservice.adapters.events;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.google.common.base.Preconditions;
import com.google.common.collect.ImmutableMap;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.MessageProperties;
import com.wks.servicemarketplace.common.messaging.Message;
import com.wks.servicemarketplace.customerservice.messaging.CustomerMessaging;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import java.io.IOException;

public class DefaultMessagePublisher {

    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultMessagePublisher.class);

    private Channel channel;

    @Inject
    public DefaultMessagePublisher(Channel amqpChannel) throws IOException {
        this.channel = amqpChannel;
        CustomerMessaging.Exchange.MAIN.declare(amqpChannel);
    }

    public boolean publish(Message message, String token) throws IOException {
        Preconditions.checkNotNull(message);


        try {
            channel.basicPublish(
                    message.getDestinationExchange(),
                    message.getDestinationRoutingKey(),
                    MessageProperties.PERSISTENT_TEXT_PLAIN.builder()
                            .correlationId(message.getCorrelationId())
                            .messageId(message.getId().toString())
                            .replyTo(message.getReplyQueue())
                            .headers(ImmutableMap.of("Authorization", String.format("Bearer %s", token)))
                            .build(),
                    message.getPayload().getBytes()
            );
            return true;
        } catch (JsonProcessingException e) {
            LOGGER.error("Failed to publish message", e);
            return false;
        }
    }
}

File: auth-service/src/main/kotlin/com/wks/servicemarketplace/authservice/adapters/events/DefaultMessagePublisher.kt
package com.wks.servicemarketplace.authservice.adapters.events

import com.rabbitmq.client.Channel
import com.rabbitmq.client.MessageProperties
import com.wks.servicemarketplace.common.messaging.Message
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import javax.inject.Inject

data class DefaultMessagePublisher @Inject constructor(private val channel: Channel) {

    companion object {
        private val LOGGER: Logger = LoggerFactory.getLogger(DefaultMessagePublisher::class.java)
    }

    fun publish(token: String, message: Message): Boolean {
        try {
            channel.basicPublish(
                    message.destinationExchange,
                    message.destinationRoutingKey,
                    MessageProperties.PERSISTENT_TEXT_PLAIN.builder()
                            .messageId(message.id.toString())
                            .replyTo(message.replyQueue)
                            .correlationId(message.correlationId)
                            .headers(mapOf(
                                    "Authorization" to "Bearer $token"
                            ))
                            .build(),
                    message.payload.toByteArray()
            )
            // Consider publisher confirms.
            return true
        } catch (e: Exception) {
            LOGGER.error("Failed to publish message '$message;", e)
            return false
        }
    }
}